// react-naver-maps/dist/hocs-65b52578.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var recompose = require('recompose');
var camelCase = _interopDefault(require('lodash.camelcase'));
require('lodash.difference');
var warning = _interopDefault(require('warning'));
var invariant = _interopDefault(require('invariant'));
var ResizeDetector = _interopDefault(require('react-resize-detector'));

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
  try {
    var info = gen[key](arg);
    var value = info.value;
  } catch (error) {
    reject(error);
    return;
  }

  if (info.done) {
    resolve(value);
  } else {
    Promise.resolve(value).then(_next, _throw);
  }
}

function _asyncToGenerator(fn) {
  return function () {
    var self = this,
        args = arguments;
    return new Promise(function (resolve, reject) {
      var gen = fn.apply(self, args);

      function _next(value) {
        asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
      }

      function _throw(err) {
        asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
      }

      _next(undefined);
    });
  };
}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};
    var ownKeys = Object.keys(source);

    if (typeof Object.getOwnPropertySymbols === 'function') {
      ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) {
        return Object.getOwnPropertyDescriptor(source, sym).enumerable;
      }));
    }

    ownKeys.forEach(function (key) {
      _defineProperty(target, key, source[key]);
    });
  }

  return target;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};

  var target = _objectWithoutPropertiesLoose(source, excluded);

  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

var pick = function pick(keys) {
  return function (props) {
    return keys.reduce(function (ret, key) {
      /* eslint-disable no-param-reassign */
      if (props[key] !== undefined) {
        ret[key] = props[key];
      }

      return ret;
      /* eslint-enable no-param-reassign */
    }, {});
  };
};

function generageEventLookup(naverEventNames) {
  return naverEventNames.reduce(function (ret, eventName) {
    // propsListener is event listener defined on props. (user input)
    var propsListenerName = camelCase("on_".concat(eventName));
    return _objectSpread({}, ret, _defineProperty({}, propsListenerName, eventName));
  }, {});
}
/**
 * Managing Naver Event Handlers. KVO instance must be registered 
 * in the child component by props.registerEventInstance
 * @param {*} WrappedComponent 
 */


var bridgeEventHandlers = function bridgeEventHandlers(WrappedComponent) {
  var Wrapper =
  /*#__PURE__*/
  function (_React$Component) {
    _inherits(Wrapper, _React$Component);

    function Wrapper(props) {
      var _this;

      _classCallCheck(this, Wrapper);

      _this = _possibleConstructorReturn(this, _getPrototypeOf(Wrapper).call(this, props));
      _this.naverListeners = {};
      _this.naverEventLookup = {};
      _this.handlingPropNames = [];
      _this.registerEventInstance = _this.registerEventInstance.bind(_assertThisInitialized(_assertThisInitialized(_this)));
      return _this;
    }

    _createClass(Wrapper, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        this.updateListeners();
      }
    }, {
      key: "componentDidUpdate",
      value: function componentDidUpdate() {
        this.updateListeners();
      }
    }, {
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        var _this2 = this;

        Object.values(this.naverListeners).forEach(function (listener) {
          _this2.unlisten(listener);
        });
      }
    }, {
      key: "getEventByHandlerName",
      value: function getEventByHandlerName(handlerName) {
        return this.naverEventLookup[handlerName];
      }
    }, {
      key: "registerEventInstance",
      value: function registerEventInstance(instance) {
        warning(!this.instance, 'react-naver-maps: bridgeEventHandlers - Tried to Change instance.');
        invariant(instance, "react-naver-maps: bridgeEventHandlers - required naver instance, but ".concat(instance));
        this.instance = instance;
        if (this.props.registerEventInstance) this.props.registerEventInstance(instance);
      }
    }, {
      key: "updateListeners",
      value: function updateListeners() {
        var _this3 = this;

        var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.props;

        if (this.eventsCache !== props.events) {
          this.eventsCache = props.events;
          this.naverEventLookup = generageEventLookup(props.events);
          this.handlingPropNames = Object.keys(this.naverEventLookup);
        } // prepare new naver listeners


        var prevNaverListeners = this.naverListeners;
        this.naverListeners = {};
        var orphans = {};
        var updateds = {};
        this.handlingPropNames.forEach(function (propName) {
          // console.log(`for ${propName}`);
          var handler = props[propName];

          if (prevNaverListeners[propName]) {
            var prevNaverListener = prevNaverListeners[propName];
            var prevHandler = prevNaverListener.listener; // handler unchanged
            // console.log(
            //   prevNaverListener,
            //   // prevHandler,
            //   // handler,
            //   prevHandler === handler,
            // );

            if (prevHandler === handler) {
              // console.log('unchanged', propName);
              _this3.naverListeners[propName] = prevNaverListeners[propName]; // handler changed
            } else {
              orphans[propName] = prevNaverListener;
              updateds[propName] = handler;
            } // new handler

          } else if (handler) {
            updateds[propName] = handler;
          }
        }); // console.log('updateds', updateds);
        // console.log('orphans', orphans);
        // listen updated handlers

        Object.keys(updateds).forEach(function (updatedPropName) {
          var evt = _this3.getEventByHandlerName(updatedPropName);

          _this3.naverListeners[updatedPropName] = _this3.listen(evt, updateds[updatedPropName]);
        }); // console.log('listeners', this.naverListeners);
        // unlisten orphan handlers

        Object.values(orphans).forEach(function (orphan) {
          _this3.unlisten(orphan);
        });
      }
    }, {
      key: "listen",
      value: function listen(eventName, listener) {
        var navermaps = this.props.navermaps;
        invariant(navermaps, 'props.navermaps required');
        invariant(this.instance, 'may be forgot to call registerEventInstance');
        return navermaps.Event.addListener(this.instance, eventName, listener);
      }
    }, {
      key: "unlisten",
      value: function unlisten(listener) {
        var navermaps = this.props.navermaps;
        navermaps.Event.removeListener(listener);
      }
    }, {
      key: "render",
      value: function render() {
        return React__default.createElement(WrappedComponent, _extends({}, this.props, {
          registerEventInstance: this.registerEventInstance
        }));
      }
    }]);

    return Wrapper;
  }(React__default.Component);

  Wrapper.displayName = recompose.wrapDisplayName(WrappedComponent, 'bridgeEventHandlers');
  return Wrapper;
};

var namedWrapper = function namedWrapper(displayName) {
  return function (WrappedComponent) {
    var Wrapper = function Wrapper(props) {
      return React__default.createElement(WrappedComponent, props);
    };

    Wrapper.displayName = displayName;
    return Wrapper;
  };
};

/**
 * inject navermaps to WrappedComponent.
 * use props.navermaps or window.naver.maps
 * if there is no naver maps module, raise.
 * @param {*} WrappedComponent
 */

var withNavermaps = function withNavermaps(WrappedComponent) {
  var Navermaps =
  /*#__PURE__*/
  function (_React$Component) {
    _inherits(Navermaps, _React$Component);

    function Navermaps(props) {
      var _this;

      _classCallCheck(this, Navermaps);

      _this = _possibleConstructorReturn(this, _getPrototypeOf(Navermaps).call(this, props));

      var navermaps = _this.getNavermaps();

      invariant(navermaps, 'props.navermaps or window.naver.maps is required.');
      return _this;
    }

    _createClass(Navermaps, [{
      key: "getNavermaps",
      value: function getNavermaps() {
        return this.props.navermaps || window.naver && window.naver.maps;
      }
    }, {
      key: "render",
      value: function render() {
        var navermaps = this.getNavermaps();
        return React__default.createElement(WrappedComponent, _extends({}, this.props, {
          navermaps: navermaps
        }));
      }
    }]);

    return Navermaps;
  }(React__default.Component);

  Navermaps.displayName = recompose.wrapDisplayName(WrappedComponent, 'withNavermaps');
  return Navermaps;
};

var injectNaverRef = function injectNaverRef(WrappedComponent) {
  function Wrapper(_ref) {
    var naverRef = _ref.naverRef,
        restProps = _objectWithoutProperties(_ref, ["naverRef"]);

    return React__default.createElement(WrappedComponent, _extends({}, restProps, {
      ref: naverRef
    }));
  }

  Wrapper.displayName = recompose.wrapDisplayName(WrappedComponent, 'injectNaverRef');
  return Wrapper;
};

var autoResize = function autoResize(WrappedComponent) {
  var Wrapper =
  /*#__PURE__*/
  function (_React$Component) {
    _inherits(Wrapper, _React$Component);

    function Wrapper(props) {
      var _this;

      _classCallCheck(this, Wrapper);

      _this = _possibleConstructorReturn(this, _getPrototypeOf(Wrapper).call(this, props));
      _this.state = {
        style: {
          width: '100%',
          height: '100%'
        }
      };
      _this.handleResize = _this.handleResize.bind(_assertThisInitialized(_assertThisInitialized(_this)));
      return _this;
    }

    _createClass(Wrapper, [{
      key: "handleResize",
      value: function handleResize(width, height) {
        this.setState({
          size: {
            width: width,
            height: height
          }
        });
      }
    }, {
      key: "render",
      value: function render() {
        var _this$props = this.props,
            id = _this$props.id,
            className = _this$props.className,
            style = _this$props.style,
            restProps = _objectWithoutProperties(_this$props, ["id", "className", "style"]);

        return React__default.createElement("div", {
          id: id,
          className: className,
          style: style
        }, React__default.createElement(WrappedComponent, _extends({}, restProps, {
          style: this.state.style,
          size: this.state.size
        })), React__default.createElement(ResizeDetector, {
          handleWidth: true,
          handleHeight: true,
          onResize: this.handleResize,
          refreshMode: "throttle",
          refreshRate: 10
        }));
      }
    }], [{
      key: "getDerivedStateFromProps",
      value: function getDerivedStateFromProps(props, state) {
        if (props.size) {
          return _objectSpread({}, state, {
            size: props.size
          });
        }

        return null;
      }
    }]);

    return Wrapper;
  }(React__default.Component);

  Wrapper.displayName = recompose.wrapDisplayName(WrappedComponent, 'autoResize');
  return Wrapper;
};

exports._extends = _extends;
exports._objectWithoutProperties = _objectWithoutProperties;
exports._objectSpread = _objectSpread;
exports._asyncToGenerator = _asyncToGenerator;
exports._classCallCheck = _classCallCheck;
exports._createClass = _createClass;
exports._possibleConstructorReturn = _possibleConstructorReturn;
exports._getPrototypeOf = _getPrototypeOf;
exports._inherits = _inherits;
exports.pick = pick;
exports.withNavermaps = withNavermaps;
exports.namedWrapper = namedWrapper;
exports.bridgeEventHandlers = bridgeEventHandlers;
exports.injectNaverRef = injectNaverRef;
exports.autoResize = autoResize;
